freqtrade/data/converter/orderflow.py

```diff
@@ -0,0 +1,277 @@
+                        if delta < min_delta:
+                            min_delta = delta
+                df.loc[is_between, "max_delta"] = max_delta
+                df.loc[is_between, "min_delta"] = min_delta
+
+                df.loc[is_between, "bid"] = np.where(
+                    trades_grouped_df["side"].str.contains("buy"), 0, trades_grouped_df["amount"]
+                ).sum()
+                df.loc[is_between, "ask"] = np.where(
+                    trades_grouped_df["side"].str.contains("sell"), 0, trades_grouped_df["amount"]
+                ).sum()
+                df.loc[is_between, "delta"] = df.loc[is_between, "ask"] - df.loc[is_between, "bid"]
+                min_delta = np.min(deltas_per_trade)
+                max_delta = np.max(deltas_per_trade)
+
+                df.loc[is_between, "total_trades"] = len(trades_grouped_df)
+                # copy to avoid memory leaks
+                dataframe.loc[is_between] = df.loc[is_between].copy()
+            else:
+                logger.debug(f"Found NO candles for trades starting with {candle_start}")
+        logger.debug(f"trades.groups_keys in {time.time() - start_time} seconds")
+
+        logger.debug(f"trades.singleton_iterate in {time.time() - start_time} seconds")
+
+    except Exception as e:
+        logger.exception("Error populating dataframe with trades:", e)
+        raise e
+
+    return dataframe
+
+
+def trades_to_volumeprofile_with_total_delta_bid_ask(trades: pd.DataFrame, scale: float):
+    """
+    :param trades: dataframe
+    :param scale: scale aka bin size e.g. 0.5
+    :return: trades binned to levels according to scale aka orderflow
+    """
+    df = pd.DataFrame([], columns=DEFAULT_ORDERFLOW_COLUMNS)
+    # create bid, ask where side is sell or buy
+    df["bid_amount"] = np.where(trades["side"].str.contains("buy"), 0, trades["amount"])
+    df["ask_amount"] = np.where(trades["side"].str.contains("sell"), 0, trades["amount"])
+    df["bid"] = np.where(trades["side"].str.contains("buy"), 0, 1)
+    df["ask"] = np.where(trades["side"].str.contains("sell"), 0, 1)
+
+    # round the prices to the nearest multiple of the scale
+    df["price"] = ((trades["price"] / scale).round() * scale).astype("float64").values
+    if df.empty:
+        df["total"] = np.nan
+        df["delta"] = np.nan
+        return df
+
+    df["delta"] = df["ask_amount"] - df["bid_amount"]
+    df["total_volume"] = df["ask_amount"] + df["bid_amount"]
+    df["total_trades"] = df["ask"] + df["bid"]
+
+    # group to bins aka apply scale
+    df = df.groupby("price").sum(numeric_only=True)
+    return df
+
+
+def trades_orderflow_to_imbalances(df: pd.DataFrame, imbalance_ratio: int, imbalance_volume: int):
+    """
+    :param df: dataframes with bid and ask
+    :param imbalance_ratio: imbalance_ratio e.g. 3
+    :param imbalance_volume: imbalance volume e.g. 10
+    :return: dataframe with bid and ask imbalance
+    """
+    bid = df.bid
+    ask = df.ask.shift(-1)
```

Comment:
This is because imbalances are compared diagonally. See https://atas.net/atas-possibilities/cluster-charts-footprint/how-to-find-and-trade-imbalance/#:~:text=Imbalance%20compares%20bids%20and%20asks,which%20is%20one%20level%20lower. for more infos

Reply:
mmmh, as said, a comment should help future readers ... as i'm sure to forget this again :laughing: